## MEIRA Code Scripts

# **Start Paper-1b**

In [ ]:
!nvidia-smi

Mon Aug  3 19:07:05 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.82.07              Driver Version: 580.82.07      CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   42C    P8             11W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [ ]:
!mkdir -p data

In [ ]:
# %%writefile run_all.sh
# #!/usr/bin/env bash
# # run_all.sh
# # ==========
# # Runs run_experiments.py -> run_ablation.py -> run_SOTA.py in sequence,
# # on the same dataset pair, with full logging and fail-fast behavior
# # (stops immediately if any step errors, so you don't burn GPU hours on
# # later steps built on a broken earlier run).
# #
# # USAGE (edit the CONFIG block below first, or override via env vars):
# #   bash run_all.sh
# #   nohup bash run_all.sh > logs/run_all.out 2>&1 &   # survive Colab disconnects
# #
# # Each step's full stdout/stderr goes to its own timestamped log in logs/,
# # in addition to the combined nohup output if you background it.

# set -euo pipefail   # -e: exit on first error, -u: error on unset vars, -o pipefail

# # ── CONFIG — edit these, or export as env vars before running ──────────
# DATASET_A="${DATASET_A:-data/topiocqa.jsonl}"
# DATASET_A_NAME="${DATASET_A_NAME:-TopiOCQA}"
# DATASET_B="${DATASET_B:-data/miracl_hi_bn.jsonl}"
# DATASET_B_NAME="${DATASET_B_NAME:-MIRACL-HiBn}"

# EXPERIMENTS_SEEDS="${EXPERIMENTS_SEEDS:-41 42 43 44 45}"
# ABLATION_SEEDS="${ABLATION_SEEDS:-41 42 43}"
# SOTA_SEEDS="${SOTA_SEEDS:-41 42 43}"
# EPOCHS="${EPOCHS:-8}"
# # ─────────────────────────────────────────────────────────────────────

# mkdir -p results logs
# TS=$(date +%Y%m%d_%H%M%S)

# echo "=================================================================="
# echo "MEIRA full pipeline run — started $(date)"
# echo "  dataset_a: $DATASET_A ($DATASET_A_NAME)"
# echo "  dataset_b: $DATASET_B ($DATASET_B_NAME)"
# echo "  epochs:    $EPOCHS"
# echo "=================================================================="

# # Sanity: fail loudly and early if the input files don't exist, rather
# # than 20 minutes into a run.
# for f in "$DATASET_A" "$DATASET_B"; do
#     if [ ! -f "$f" ]; then
#         echo "[run_all] ERROR: dataset file not found: $f"
#         echo "[run_all] Run the appropriate prepare_*.py script first."
#         exit 1
#     fi
# done

# echo ""
# echo "── [1/3] run_experiments.py ─────────────────────────────────────"
# python run_experiments.py \
#     --dataset_a "$DATASET_A" --dataset_a_name "$DATASET_A_NAME" \
#     --dataset_b "$DATASET_B" --dataset_b_name "$DATASET_B_NAME" \
#     --seeds $EXPERIMENTS_SEEDS --epochs "$EPOCHS" \
#     --output results/experiments.json \
#     2>&1 | tee "logs/experiments_${TS}.log"

# echo ""
# echo "── [2/3] run_ablation.py ────────────────────────────────────────"
# python run_ablation.py \
#     --dataset_a "$DATASET_A" --dataset_a_name "$DATASET_A_NAME" \
#     --dataset_b "$DATASET_B" --dataset_b_name "$DATASET_B_NAME" \
#     --seeds $ABLATION_SEEDS --epochs "$EPOCHS" --significance \
#     --output results/ablation.json \
#     2>&1 | tee "logs/ablation_${TS}.log"

# echo ""
# echo "── [3/3] run_SOTA.py ────────────────────────────────────────────"
# python run_SOTA.py \
#     --dataset_a "$DATASET_A" --dataset_a_name "$DATASET_A_NAME" \
#     --dataset_b "$DATASET_B" --dataset_b_name "$DATASET_B_NAME" \
#     --seeds $SOTA_SEEDS \
#     --output results/sota_comparison.json \
#     2>&1 | tee "logs/sota_${TS}.log"

# echo ""
# echo "=================================================================="
# echo "All three runs completed successfully: $(date)"
# echo "  results/experiments.json"
# echo "  results/ablation.json"
# echo "  results/sota_comparison.json"
# echo "  full logs in logs/*_${TS}.log"
# echo "=================================================================="

In [ ]:
!bash run_all.sh
#!nohup bash run_all.sh > logs/run_all.out 2>&1 &

In [ ]:
!nvidia-smi

Mon Aug  3 14:29:52 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.82.07              Driver Version: 580.82.07      CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   50C    P8             12W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [ ]:
!pip install -q "datasets<4.0.0"

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 491.5/491.5 kB 27.2 MB/s eta 0:00:00


In [ ]:
!mkdir -p data

In [ ]:
!python prepare_topiocqa.py --split train --out data/topiocqa.jsonl --neg_per_pos 3

[prepare_topiocqa] downloading McGill-NLP/TopiOCQA ...
README.md: 100% 2.17k/2.17k [00:00<00:00, 6.41MB/s]
TopiOCQA.py: 100% 4.57k/4.57k [00:00<00:00, 12.0MB/s]

plain_text/train/0000.parquet: downloading bytes:  90% 25.9M/28.9M [00:06<00:00, 13.2MB/s, 1.86MB/s  ]
plain_text/train/0000.parquet: downloading bytes: 100% 28.8M/28.8M [00:06<00:00, 4.49MB/s, 2.39MB/s  ]
plain_text/train/0000.parquet: reconstructing file: 100% 28.9M/28.9M [00:06<00:00, 4.51MB/s, 2.53MB/s  ]

plain_text/validation/0000.parquet: downloading bytes:   0% 0.00/2.64M [00:00<?, ?B/s]
plain_text/validation/0000.parquet: downloading bytes: 100% 2.63M/2.63M [00:02<00:00, 1.29MB/s,  251kB/s  ]
plain_text/validation/0000.parquet: reconstructing file: 100% 2.64M/2.64M [00:02<00:00, 1.30MB/s,  252kB/s  ]
Generating train split: 100% 45450/45450 [00:00<00:00, 138305.86 examples/s]
Generating validation split: 100% 2514/2514 [00:00<00:00, 109319.07 examples/s]

[prepare_topiocqa] RAW SCHEMA CHECK — first example:
{
  "Conve

In [ ]:
!python validate_jsonl.py data/topiocqa.jsonl


=== data/topiocqa.jsonl ===
lines read:        181792
valid rows:        181792
unique qids:       45450
unique conv_ids:   3509
label distribution:{1: 45450, 0: 136342}
  -> 45450/181792 = 25.0% positive
  ✅ File is well-formed and ready for meira/data.py's load_jsonl().


In [ ]:
!cat data/miracl_hi.jsonl data/miracl_bn.jsonl > data/miracl_hi_bn.jsonl
!python validate_jsonl.py data/miracl_hi_bn.jsonl

cat: data/miracl_hi.jsonl: No such file or directory
cat: data/miracl_bn.jsonl: No such file or directory

=== data/miracl_hi_bn.jsonl ===
lines read:        0
valid rows:        0
unique qids:       0
unique conv_ids:   0
label distribution:{}


In [ ]:
!python run_experiments.py \
    --dataset_a data/topiocqa.jsonl --dataset_a_name TopiOCQA \
    --dataset_b data/miracl_hi_bn.jsonl --dataset_b_name MIRACL-HiBn \
    --seeds 41 42 43 44 45 --epochs 8 \
    --output results/experiments.json

Traceback (most recent call last):
  File "/content/run_experiments.py", line 33, in <module>
    from meira.cli_common import add_dataset_args, load_datasets, build_vocab_from_train, \
ModuleNotFoundError: No module named 'meira'


In [ ]:
!python run_ablation.py \
    --dataset_a data/topiocqa.jsonl --dataset_a_name TopiOCQA \
    --dataset_b data/miracl_hi_bn.jsonl --dataset_b_name MIRACL-HiBn \
    --seeds 41 42 43 --epochs 8 --significance \
    --output results/ablation.json

In [ ]:
!python run_SOTA.py \
    --dataset_a data/topiocqa.jsonl --dataset_a_name TopiOCQA \
    --dataset_b data/miracl_hi_bn.jsonl --dataset_b_name MIRACL-HiBn \
    --seeds 41 42 43 \
    --output results/sota_comparison.json

# **Start Paper-1**

In [ ]:
!unzip -q '/content/MEIRA_FIRE2026_ACM_suite.zip'

In [ ]:
%cd '/content/meira_fire2026'

/content/meira_fire2026


In [ ]:
!python run_SOTA.py


  MEIRA – FIRE 2026 SOTA Comparison
  Leaderboard on Two IR Benchmark Datasets

[1/3] Building datasets
  FIRE-AgentIR-2026: 8,400 samples  pos=2302(27%)  hard_neg=4094
  FIRE-CrossLingIR-2026: 3,200 samples  pos=920(29%)  hard_neg=1457

[2/3] Running SOTA comparison (5 seeds × 5 models)

  Dataset: FIRE-AgentIR-2026
    BM25           F1=0.490  nDCG@10=0.812  MAP=0.736  MRR=0.511
    TF-IDF         F1=0.479  nDCG@10=0.809  MAP=0.731  MRR=0.510
    Dense-IR       F1=0.648  nDCG@10=0.902  MAP=0.859  MRR=0.593
    ColBERT-like   F1=0.740  nDCG@10=0.940  MAP=0.912  MRR=0.624
    MEIRA-full     F1=0.826  nDCG@10=0.970  MAP=0.957  MRR=0.650  XAIR@10=0.896  MDS=1.000

  Dataset: FIRE-CrossLingIR-2026
    BM25           F1=0.465  nDCG@10=0.815  MAP=0.745  MRR=0.421
    TF-IDF         F1=0.455  nDCG@10=0.811  MAP=0.740  MRR=0.419
    Dense-IR       F1=0.588  nDCG@10=0.890  MAP=0.847  MRR=0.478
    ColBERT-like   F1=0.674  nDCG@10=0.926  MAP=0.896  MRR=0.504
    MEIRA-full     F1=0.778  nDCG@1

In [ ]:
!python run_ablation.py


  MEIRA – FIRE 2026 Ablation Study
  Component Contribution on Two IR Benchmark Datasets

[1/3] Building datasets
  FIRE-AgentIR-2026: 8,400 samples  pos=2302(27%)  hard_neg=4094
  FIRE-CrossLingIR-2026: 3,200 samples  pos=920(29%)  hard_neg=1457

[2/3] Running ablation (5 seeds × 4 variants)

  Dataset: FIRE-AgentIR-2026
    MEIRA-full         (Full model (memory + XAI + decay))  F1=0.826±0.011  nDCG@10=0.970  XAIR@10=0.896  MDS=1.000
    MEIRA-no-memory    (− Episodic memory               )  F1=0.716±0.014  nDCG@10=0.933  XAIR@10=0.861  MDS=0.000
    MEIRA-no-xai       (− XAI attribution head          )  F1=0.779±0.009  nDCG@10=0.955  XAIR@10=0.000  MDS=1.000
    MEIRA-no-decay     (− Temporal decay                )  F1=0.748±0.011  nDCG@10=0.944  XAIR@10=0.872  MDS=1.000

  Dataset: FIRE-CrossLingIR-2026
    MEIRA-full         (Full model (memory + XAI + decay))  F1=0.778±0.025  nDCG@10=0.956  XAIR@10=0.882  MDS=1.000
    MEIRA-no-memory    (− Episodic memory               )  F1=0.

In [ ]:
!python run_experiments.py


  MEIRA – FIRE 2026 Experiments
  K-Fold + Multi-Seed on Two IR Benchmark Datasets

[1/4] Building datasets
  FIRE-AgentIR-2026: 8,400 samples  pos=2302(27%)  hard_neg=4094
  FIRE-CrossLingIR-2026: 3,200 samples  pos=920(29%)  hard_neg=1457

[2/4] K-Fold (k=5) – MEIRA-full

  K-Fold (k=5) – MEIRA-full
    [FIRE-AgentIR-2026] Fold 1/5  F1=0.827  nDCG@10=0.965  XAIR@10=0.892  MDS=1.000
    [FIRE-AgentIR-2026] Fold 2/5  F1=0.832  nDCG@10=0.972  XAIR@10=0.899  MDS=1.000
    [FIRE-AgentIR-2026] Fold 3/5  F1=0.816  nDCG@10=0.966  XAIR@10=0.893  MDS=1.000
    [FIRE-AgentIR-2026] Fold 4/5  F1=0.822  nDCG@10=0.956  XAIR@10=0.882  MDS=1.000
    [FIRE-AgentIR-2026] Fold 5/5  F1=0.824  nDCG@10=0.970  XAIR@10=0.894  MDS=1.000
    [FIRE-CrossLingIR-2026] Fold 1/5  F1=0.754  nDCG@10=0.949  XAIR@10=0.877  MDS=1.000
    [FIRE-CrossLingIR-2026] Fold 2/5  F1=0.770  nDCG@10=0.948  XAIR@10=0.876  MDS=1.000
    [FIRE-CrossLingIR-2026] Fold 3/5  F1=0.794  nDCG@10=0.950  XAIR@10=0.871  MDS=1.000
    [FIRE-Cr

# Understanding the MEIRA Pipeline: A Step-by-Step Explanation

The MEIRA (Memory-Enhanced Information Retrieval Agent) pipeline is designed for advanced information retrieval tasks, particularly focusing on conversational and temporal aspects. It combines transformer architectures with novel memory and time-awareness mechanisms to improve retrieval performance.

This explanation will walk through the core components and scripts found in the `/content/meira_fire2026` directory.

## Step 1: Data Preparation

The data preparation phase is crucial for transforming raw information into a format suitable for the MEIRA model. The pipeline primarily uses JSONL files, where each line represents a data record (e.g., a query-document pair with a label). The `meira/data.py` script is central to this process, often supported by `prepare_*.py` scripts for specific datasets.

### Core Concepts:

*   **JSONL Format:** Data is expected in JSON Lines format, with each line being a JSON object. Key fields include `qid` (query ID), `conv_id` (conversation ID), `turn` (turn number in a conversation), `query`, `doc` (document text), and `label` (relevance label, e.g., 0 for irrelevant, 1 for relevant).
*   **Vocabulary Creation:** The system builds a vocabulary of unique tokens from the training data. This maps each unique word or subword unit to an integer index, which the embedding layers of the model use.
*   **Sequence Preparation:** Input data (queries and documents) are converted into sequences of token IDs. The `seq_len` (sequence length) parameter defines the maximum length of these sequences. If a sequence is shorter, it's padded; if longer, it's truncated.
*   **Timestamp Normalization:** Timestamps associated with tokens or events are normalized (e.g., to days from a common epoch) to allow the model to incorporate temporal information.
*   **Dataset Splitting:** The data is typically split into training, validation, and test sets. This ensures the model is evaluated on unseen data to prevent overfitting. For robust evaluation, K-fold cross-validation or multi-seed splitting might be used.

### Example (Conceptual `meira/data.py` operations):

```python
# pseudo-code to illustrate data loading and tokenization
import json
import torch
from collections import defaultdict

class Dataset:
    def __init__(self, data_path, seq_len, max_days_between_tokens):
        self.data_path = data_path
        self.seq_len = seq_len
        self.max_days_between_tokens = max_days_between_tokens
        self.token_to_idx = {}
        self.idx_to_token = []
        self.num_tokens = 0
        self.sequences = []

        self._load_data()
        self._build_vocab() # Simplified: usually involves pre-trained tokenizers or custom building
        self._prepare_sequences()

    def _load_data(self):
        with open(self.data_path, 'r') as f:
            raw_data = [json.loads(line) for line in f]
        # Further processing like timestamp normalization happens here
        self.data = raw_data # Store processed data

    def _build_vocab(self): # Highly simplified
        unique_tokens = set()
        for record in self.data:
            # Assume a simple split for demonstration; real-world uses tokenizers
            for word in record['query'].split() + record['doc'].split():
                unique_tokens.add(word)
        self.idx_to_token = sorted(list(unique_tokens))
        self.token_to_idx = {token: idx for idx, token in enumerate(self.idx_to_token)}
        self.num_tokens = len(self.idx_to_token)

    def _prepare_sequences(self):
        for record in self.data:
            query_tokens = [self.token_to_idx.get(w, 0) for w in record['query'].split()][:self.seq_len]
            doc_tokens = [self.token_to_idx.get(w, 0) for w in record['doc'].split()][:self.seq_len]

            # Example sequence structure, actual one is more complex with target token, timestamps, etc.
            self.sequences.append({
                'input_tokens_query': torch.tensor(query_tokens, dtype=torch.long),
                'input_tokens_doc': torch.tensor(doc_tokens, dtype=torch.long),
                'label': torch.tensor(record['label'], dtype=torch.long),
                'timestamp': torch.tensor(record['timestamp'], dtype=torch.float)
            })

    def __len__(self):
        return len(self.sequences)

    def __getitem__(self, idx):
        return self.sequences[idx]

# The actual `meira/data.py` (or `dataloader.py` in v2) is more sophisticated, handling specific splitting and creating 'hard negatives' for training.

## Step 2: MEIRA Model Architecture (`meira/model.py`)

The `meira/model.py` script defines the neural network architecture of MEIRA, which is a transformer-based model enhanced with components specifically designed for temporal and memory-aware processing. It consists of several key modules:

### 2.1. Positional Encoding

Like many transformer models, MEIRA uses positional encoding to inject information about the relative or absolute position of tokens in a sequence. Since transformers process input sequences in parallel without inherent order, positional encodings are added to the token embeddings to give the model a sense of position.

*   **Mechanism:** Uses sine and cosine functions of different frequencies to create unique positional vectors. These vectors are added to the word embeddings.
*   **File:** Defined in `meira/model.py`.

```python
# From meira/model.py
class PositionalEncoding(nn.Module):
    def __init__(self, d_model, max_len=5000):
        super().__init__()
        pe = torch.zeros(max_len, d_model)
        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
        div_term = torch.exp(torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model))
        pe[:, 0::2] = torch.sin(position * div_term)
        pe[:, 1::2] = torch.cos(position * div_term)
        pe = pe.unsqueeze(0).transpose(0, 1)
        self.register_buffer('pe', pe)

    def forward(self, x):
        return x + self.pe[:x.size(0), :]
```

### 2.2. Time-Aware Attention

This is a custom attention mechanism that extends standard multi-head attention by incorporating temporal information. It modifies the attention scores based on the time difference between tokens in a sequence.

*   **Mechanism:** Learnable parameters (`time_w`, `time_b`) are used within a sigmoid function to create a time-encoding factor. This factor then scales the standard attention scores, effectively giving more or less weight to tokens based on their temporal proximity or distance.
*   **File:** Defined in `meira/model.py`.

```python
# From meira/model.py (simplified relevant part)
class TimeAwareAttention(nn.Module):
    # ... initialization ...

    def forward(self, query, key, value, timestamps, mask=None):
        # ... standard attention score calculation ...
        scores = torch.matmul(Q, K.permute(0, 1, 3, 2)) / self.scale

        # Add time awareness
        time_diff = timestamps.unsqueeze(1).unsqueeze(1) - timestamps.unsqueeze(2).unsqueeze(1)
        time_encoding = torch.sigmoid(self.time_w * time_diff + self.time_b)
        scores = scores * time_encoding # Apply time encoding to attention scores

        # ... softmax, dropout, output projection ...
        return output
```

### 2.3. Memory Module (Episodic Memory)

This module is designed to store and retrieve long-term episodic memory. It maintains learnable representations (keys, values, and age) for each token, which are updated based on the input sequence and its timestamp.

*   **Mechanism:**
    *   `memory_keys`: Learnable embeddings for each token, used to match against input embeddings.
    *   `memory_values`: Learnable representations for each token, which store historical context.
    *   `memory_age`: Learnable initial age for each token, influencing time decay.
    *   **Time-Decay:** When retrieving from memory, values are decayed based on the time difference between the current input's timestamp and the stored memory's age (`time_decay = exp(-|current_timestamp - memory_age|)`).
    *   **Memory Update:** Input embeddings are combined with time-decayed memory values to update the memory representation for subsequent processing.
    *   **Regularization:** A regularization loss (`lambda_mem * mean(norm(keys - values))`) encourages similar tokens to have similar memory states, promoting consistency.
*   **File:** Defined in `meira/model.py`.

```python
# From meira/model.py (simplified relevant part)
class MemoryModule(nn.Module):
    # ... initialization ...

    def forward(self, input_embeddings, input_tokens, input_timestamps):
        keys = self.memory_keys[input_tokens]
        values = self.memory_values[input_tokens]
        ages = self.memory_age[input_tokens]

        # Calculate time-decayed weights
        time_decay = torch.exp(-torch.abs(input_timestamps - ages.unsqueeze(-1)))
        decayed_values = values * time_decay

        # Update memory values
        updated_memory = self.update_layer(torch.cat([input_embeddings, decayed_values], dim=-1))

        # Regularization loss
        mem_reg_loss = self.lambda_mem * torch.mean(torch.norm(keys - values, p=2, dim=-1))

        return updated_memory, mem_reg_loss
```

### 2.4. MEIRA (Overall Model)

The main `MEIRA` class integrates all these components.

*   **Workflow:**
    1.  Input tokens are embedded and combined with positional encodings.
    2.  The resulting embeddings are passed through the `MemoryModule` to retrieve and update memory-enhanced representations.
    3.  These memory-enhanced representations then go through a standard `TransformerEncoder` (which may internally use or be influenced by the `TimeAwareAttention` depending on the exact implementation details of the transformer layers).
    4.  Finally, an output linear layer (`fc_out`) predicts the relevance or category based on the transformer's output for the last token.
*   **File:** Defined in `meira/model.py`.

```python
# From meira/model.py (simplified relevant part)
class MEIRA(nn.Module):
    # ... initialization (token_embedding, positional_encoding, transformer_encoder, memory_module, fc_out) ...

    def forward(self, input_tokens, input_timestamps):
        token_embeddings = self.token_embedding(input_tokens)
        embedded_sequences = self.positional_encoding(token_embeddings)

        memory_output, mem_reg_loss = self.memory_module(embedded_sequences, input_tokens, input_timestamps)

        transformer_output = self.transformer_encoder(memory_output)

        output = self.fc_out(transformer_output[:, -1, :])

        return output, mem_reg_loss
```

## Step 3: Training and Evaluation (`meira/engine.py` or `meira_code/train.py`)

The training and evaluation logic is encapsulated in the `Trainer` class (found in `meira/engine.py` or `meira_code/train.py`). This component manages the training loop, calculates losses, performs optimization, and evaluates the model's performance on validation and test sets.

### Core Functions:

*   **Optimizer:** Uses `Adam` for updating model weights.
*   **Loss Function:** `CrossEntropyLoss` is typically used for classification tasks. The total loss combines the task-specific cross-entropy loss with a memory regularization loss (`lambda_aux * mem_reg_loss`) to encourage meaningful memory states.
*   **Learning Rate Scheduler:** `StepLR` adjusts the learning rate during training, often decaying it at specified intervals to help with convergence.
*   **Training Loop:** Iterates through epochs, processing data in batches:
    *   Performs a forward pass through the MEIRA model.
    *   Calculates the combined loss.
    *   Performs backpropagation (`loss.backward()`) and updates weights (`optimizer.step()`).
*   **Evaluation (`_evaluate` method):** After each epoch (or periodically), the model is switched to evaluation mode (`model.eval()`). It then computes various metrics on the validation set:
    *   **Classification Metrics:** F1-score, AUC (Area Under the Receiver Operating Characteristic curve), Precision, Recall, Accuracy.
    *   **Ranking Metrics (for IR):** nDCG (Normalized Discounted Cumulative Gain), MAP (Mean Average Precision), MRR (Mean Reciprocal Rank), P@K (Precision at K).
    *   **Novelty Metrics (Specific to MEIRA):** XAIR (eXplainable Artificial Intelligence for Information Retrieval), MDS (Memory Disentanglement Score) - these are likely custom metrics to assess the interpretability and effectiveness of the memory module.
*   **Early Stopping:** To prevent overfitting and save computational resources, training stops if the validation performance (e.g., F1-score) does not improve for a certain number of epochs (`patience`).
*   **Checkpointing:** The best performing model (based on validation metrics) is saved (`best.pt`).

### Example (Conceptual `Trainer` methods):

```python
# From meira_code/train.py (simplified relevant parts)
class Trainer:
    # ... initialization (model, dataset, device, optimizer, criterion, scheduler, etc.) ...

    def train(self):
        best_val_f1 = -1.0
        epochs_no_improve = 0
        for epoch in range(self.num_epochs):
            self.model.train()
            total_loss = 0.0
            for batch in self.train_loader:
                # ... load input_tokens, input_timestamps, target_token to device ...
                self.optimizer.zero_grad()
                outputs, mem_reg_loss = self.model(input_tokens, input_timestamps)
                task_loss = self.criterion(outputs, target_token)
                loss = task_loss + self.lambda_aux * mem_reg_loss # Combined loss
                loss.backward()
                self.optimizer.step()
                total_loss += loss.item()
            avg_train_loss = total_loss / len(self.train_loader)

            val_f1, val_loss = self._evaluate(self.val_loader, calculate_loss=True)
            print(f'  Ep {epoch+1:02d} | T-Loss:{avg_train_loss:.4f} | V-F1:{val_f1:.3f}')
            self.scheduler.step()

            if val_f1 > best_val_f1:
                best_val_f1 = val_f1
                epochs_no_improve = 0
                torch.save(self.model.state_dict(), os.path.join(self.checkpoints_dir, 'best.pt'))
            else:
                epochs_no_improve += 1
                if epochs_no_improve == self.patience:
                    print(f'Early stopping after {self.patience} epochs without improvement.')
                    break

    def _evaluate(self, data_loader, calculate_loss=False):
        self.model.eval()
        # ... calculate predictions, targets, and various metrics (F1, AUC, etc.) ...
        return f1_score, avg_loss # Simplified
```

## Step 4: Experiment Execution and Analysis

The pipeline includes several scripts (`run_experiments.py`, `run_ablation.py`, `run_SOTA.py`) to systematically evaluate MEIRA under different conditions and compare it with baselines.

### 4.1. General Experiments (`run_experiments.py`)

This script is used for running standard experiments to establish the model's performance. It supports:

*   **Multi-Seed Runs:** Running the model multiple times with different random seeds (e.g., `seeds 41 42 43 44 45`) to assess the stability and average performance.
*   **K-Fold Cross-Validation:** Dividing the dataset into `K` folds, training on `K-1` folds, and validating on the remaining fold, rotating until each fold has been used for validation. This provides a more robust estimate of performance, especially on smaller datasets.
*   **Output:** Generates a `results/experiments.json` file containing all measured metrics and plots (e.g., `exp1_kfold_boxplot.png` for K-fold results, `exp2_seed_variance.png` for visualizing variance across seeds).

### 4.2. Ablation Studies (`run_ablation.py`)

Ablation studies are crucial for understanding the contribution of individual components within the MEIRA model. This script systematically removes or modifies key components and measures the impact on performance.

*   **Variants:** Common ablation variants might include:
    *   `MEIRA-full`: The complete model.
    *   `MEIRA-no-memory`: Model without the episodic memory module.
    *   `MEIRA-no-xai`: Model without the XAI attribution head (if present).
    *   `MEIRA-no-decay`: Model with time-aware attention but without the temporal decay mechanism.
*   **Significance Testing:** The `--significance` flag is often used to perform statistical tests (e.g., t-tests) to determine if the performance differences between the full model and ablated variants are statistically significant.
*   **Output:** Produces `results/ablation.json` and figures (e.g., `abl1_component_bars.png`, `abl2_delta_heatmap.png`) showing the performance drop for each ablated component.

### 4.3. State-of-the-Art (SOTA) Comparison (`run_SOTA.py`)

This script compares the MEIRA model's performance against various baseline and state-of-the-art information retrieval models.

*   **Baselines:** Common baselines include:
    *   **BM25 (Best Match 25):** A classical statistical retrieval model.
    *   **TF-IDF (Term Frequency-Inverse Document Frequency):** Another classical model based on term weighting.
    *   **Dense-IR:** Models using dense vector representations (e.g., from neural embeddings) for retrieval.
    *   **ColBERT-like:** Models inspired by ColBERT, which use late interaction for efficient and effective retrieval.
*   **Evaluation:** The script trains and evaluates these baseline models alongside MEIRA on the same datasets, providing a comprehensive leaderboard.
*   **Significance:** Similar to ablation studies, it can perform significance tests to determine if MEIRA's performance improvement over baselines is statistically significant.
*   **Output:** Generates `results/sota_comparison.json` and figures (e.g., `sota1_leaderboard_bars.png`, `sota2_novel_metrics.png`) to visualize the comparative performance.

## Conclusion

In summary, the MEIRA pipeline provides a robust framework for developing and evaluating advanced information retrieval models. It meticulously handles data preparation, employs a sophisticated memory-enhanced transformer architecture, and includes comprehensive tools for training, evaluating, and comparing models through various experimental setups. This systematic approach allows for a deep understanding of the model's capabilities and the contribution of its unique components, particularly in handling temporal and conversational aspects of information retrieval.

## Summary of Experimental Results

Based on the execution outputs from `run_experiments.py`, `run_ablation.py`, and `run_SOTA.py`, we can summarize the performance of the MEIRA model and its components across two key datasets: FIRE-AgentIR-2026 and FIRE-CrossLingIR-2026.

### 1. Overall MEIRA-full Performance (from `run_experiments.py`)

The MEIRA-full model demonstrates strong performance across various metrics, especially in nDCG@10, XAIR@10, and MDS, which are crucial for information retrieval and explainability.

| Metric        | FIRE-AgentIR-2026 (Mean ± Std) | FIRE-CrossLingIR-2026 (Mean ± Std) |
| :------------ | :----------------------------- | :--------------------------------- |
| **F1**        | 0.826 ± 0.011                  | 0.778 ± 0.025                      |
| **AUC**       | 0.956 ± 0.004                  | 0.924 ± 0.012                      |
| **nDCG@10**   | 0.970 ± 0.004                  | 0.956 ± 0.006                      |
| **MAP**       | 0.957 ± 0.006                  | 0.939 ± 0.008                      |
| **MRR**       | 0.650 ± 0.011                  | 0.525 ± 0.009                      |
| **XAIR@10**   | 0.896 ± 0.004                  | 0.882 ± 0.006                      |
| **MDS**       | 1.000 ± 0.000                  | 1.000 ± 0.000                      |

### 2. Ablation Study Results (from `run_ablation.py`)

This study highlights the contribution of each core MEIRA component. The metrics shown are for the mean performance across 5 seeds.

#### FIRE-AgentIR-2026

| Variant         | F1 Score | nDCG@10 | XAIR@10 | MDS    |
| :-------------- | :------- | :------ | :------ | :----- |
| **MEIRA-full**  | 0.826    | 0.970   | 0.896   | 1.000  |
| MEIRA-no-memory | 0.716    | 0.933   | 0.861   | 0.000  |
| MEIRA-no-xai    | 0.779    | 0.955   | 0.000   | 1.000  |
| MEIRA-no-decay  | 0.748    | 0.944   | 0.872   | 1.000  |

**Component Contribution (Δ = MEIRA-full - Variant):**

| Component Ablated   | ΔF1     | ΔnDCG@10 | ΔXAIR@10 | ΔMDS   |
| :------------------ | :------ | :------- | :------- | :----- |
| Episodic memory     | +0.110  | +0.038   | +0.035   | +1.000 |
| XAI attribution head| +0.046  | +0.016   | +0.896   | +0.000 |
| Temporal decay      | +0.078  | +0.027   | +0.024   | +0.000 |

#### FIRE-CrossLingIR-2026

| Variant         | F1 Score | nDCG@10 | XAIR@10 | MDS    |
| :-------------- | :------- | :------ | :------ | :----- |
| **MEIRA-full**  | 0.778    | 0.956   | 0.882   | 1.000  |
| MEIRA-no-memory | 0.648    | 0.916   | 0.846   | 0.000  |
| MEIRA-no-xai    | 0.723    | 0.941   | 0.000   | 1.000  |
| MEIRA-no-decay  | 0.689    | 0.931   | 0.863   | 1.000  |

**Component Contribution (Δ = MEIRA-full - Variant):**

| Component Ablated   | ΔF1     | ΔnDCG@10 | ΔXAIR@10 | ΔMDS   |
| :------------------ | :------ | :------- | :------- | :----- |
| Episodic memory     | +0.130  | +0.040   | +0.036   | +1.000 |
| XAI attribution head| +0.055  | +0.015   | +0.882   | +0.000 |
| Temporal decay      | +0.089  | +0.025   | +0.019   | +0.000 |

These results clearly indicate that the episodic memory module is the most significant contributor to the MEIRA model's F1 score and nDCG@10, while the XAI attribution head has the largest impact on the XAIR@10 metric, as expected.

### 3. State-of-the-Art (SOTA) Comparison (from `run_SOTA.py`)

MEIRA-full consistently outperforms classical and neural baselines on both datasets.

#### FIRE-AgentIR-2026 Leaderboard

| Model          | F1 Score | nDCG@10 | MAP   | MRR   | XAIR@10 | MDS    |
| :------------- | :------- | :------ | :---- | :---- | :------ | :----- |
| BM25           | 0.490    | 0.812   | 0.736 | 0.511 | -       | -      |
| TF-IDF         | 0.479    | 0.809   | 0.731 | 0.510 | -       | -      |
| Dense-IR       | 0.648    | 0.902   | 0.859 | 0.593 | -       | -      |
| ColBERT-like   | 0.740    | 0.940   | 0.912 | 0.624 | -       | -      |
| **MEIRA-full** | **0.826**| **0.970**| **0.957**| **0.650**| **0.896**| **1.000**|

#### FIRE-CrossLingIR-2026 Leaderboard

| Model          | F1 Score | nDCG@10 | MAP   | MRR   | XAIR@10 | MDS    |
| :------------- | :------- | :------ | :---- | :---- | :------ | :----- |
| BM25           | 0.465    | 0.815   | 0.745 | 0.421 | -       | -      |
| TF-IDF         | 0.455    | 0.811   | 0.740 | 0.419 | -       | -      |
| Dense-IR       | 0.588    | 0.890   | 0.847 | 0.478 | -       | -      |
| ColBERT-like   | 0.674    | 0.926   | 0.896 | 0.504 | -       | -      |
| **MEIRA-full** | **0.778**| **0.956**| **0.939**| **0.525**| **0.882**| **1.000**|

MEIRA-full shows a statistically significant improvement in nDCG@10 over the best baseline (ColBERT-like) on both FIRE-AgentIR-2026 (p=0.0001) and FIRE-CrossLingIR-2026 (p=0.0008), demonstrating its effectiveness in achieving superior retrieval performance.

### **End of Paper-1**

# **Start of Paper-1a**

In [ ]:
!unzip -q '/content/MEIRA_FIRE2026_COMPLETE.zip'

In [ ]:
!python src/run_experiments.py --k 5 --seeds 5


  MEIRA – Experiments (K-Fold + Multi-Seed)
  FIRE 2026 | Two IR Benchmark Datasets

[1/4] Building datasets
  FIRE-AgentIR-2026: 8,400  pos=2302(27%)  hard_neg=4094
  FIRE-CrossLingIR-2026: 3,200  pos=920(29%)  hard_neg=1457

[2/4] K-Fold (k=5)
    [FIRE-AgentIR-2] Fold 1/5  F1=0.827  nDCG@10=0.965  XAIR@10=0.892  MDS=1.000
    [FIRE-AgentIR-2] Fold 2/5  F1=0.832  nDCG@10=0.972  XAIR@10=0.899  MDS=1.000
    [FIRE-AgentIR-2] Fold 3/5  F1=0.816  nDCG@10=0.966  XAIR@10=0.893  MDS=1.000
    [FIRE-AgentIR-2] Fold 4/5  F1=0.822  nDCG@10=0.956  XAIR@10=0.882  MDS=1.000
    [FIRE-AgentIR-2] Fold 5/5  F1=0.824  nDCG@10=0.970  XAIR@10=0.894  MDS=1.000
    [FIRE-CrossLing] Fold 1/5  F1=0.754  nDCG@10=0.949  XAIR@10=0.877  MDS=1.000
    [FIRE-CrossLing] Fold 2/5  F1=0.770  nDCG@10=0.948  XAIR@10=0.876  MDS=1.000
    [FIRE-CrossLing] Fold 3/5  F1=0.794  nDCG@10=0.950  XAIR@10=0.871  MDS=1.000
    [FIRE-CrossLing] Fold 4/5  F1=0.767  nDCG@10=0.959  XAIR@10=0.881  MDS=1.000
    [FIRE-CrossLing] Fol

In [ ]:
!python src/run_ablation.py    --seeds 5


  MEIRA Ablation Study – FIRE 2026

[1/3] Building datasets
  FIRE-AgentIR-2026: 8,400 samples  pos=2302  neg=6098
  FIRE-CrossLingIR-2026: 3,200 samples  pos=920  neg=2280

[2/3] Evaluating 8 models × 5 seeds × 2 datasets
  [FIRE-AgentIR] BM25                   F1=0.490±0.017  nDCG@10=0.812  XAIR@10=0.000  MDS=0.000
  [FIRE-AgentIR] TF-IDF                 F1=0.479±0.019  nDCG@10=0.809  XAIR@10=0.000  MDS=0.000
  [FIRE-AgentIR] Dense-IR               F1=0.648±0.018  nDCG@10=0.902  XAIR@10=0.000  MDS=0.000
  [FIRE-AgentIR] ColBERT-like           F1=0.740±0.013  nDCG@10=0.940  XAIR@10=0.000  MDS=0.000
  [FIRE-AgentIR] MEIRA-no-memory        F1=0.716±0.014  nDCG@10=0.933  XAIR@10=0.861  MDS=0.000
  [FIRE-AgentIR] MEIRA-no-decay         F1=0.748±0.011  nDCG@10=0.944  XAIR@10=0.872  MDS=1.000
  [FIRE-AgentIR] MEIRA-no-xai           F1=0.779±0.009  nDCG@10=0.955  XAIR@10=0.000  MDS=1.000
  [FIRE-AgentIR] MEIRA-full             F1=0.826±0.011  nDCG@10=0.970  XAIR@10=0.896  MDS=1.000
  [FIRE-

In [ ]:
!python src/run_SOTA.py --seeds 5


  MEIRA vs SOTA – FIRE 2026

[1/3] Building datasets
  FIRE-AgentIR-2026: 8,400 samples  pos=2302  neg=6098
  FIRE-CrossLingIR-2026: 3,200 samples  pos=920  neg=2280

[2/3] Evaluating 10 SOTA models × 5 seeds × 2 datasets
  [FIRE-AgentIR] BM25             F1=0.490±0.017  nDCG@10=0.812  XAIR@10=0.000
  [FIRE-AgentIR] TF-IDF           F1=0.479±0.019  nDCG@10=0.809  XAIR@10=0.000
  [FIRE-AgentIR] DPR              F1=0.636±0.020  nDCG@10=0.896  XAIR@10=0.000
  [FIRE-AgentIR] ANCE             F1=0.698±0.015  nDCG@10=0.927  XAIR@10=0.000
  [FIRE-AgentIR] ColBERT          F1=0.740±0.013  nDCG@10=0.940  XAIR@10=0.000
  [FIRE-AgentIR] RAG-Token        F1=0.676±0.018  nDCG@10=0.917  XAIR@10=0.000
  [FIRE-AgentIR] RAG-Seq          F1=0.698±0.015  nDCG@10=0.927  XAIR@10=0.000
  [FIRE-AgentIR] MemGPT-IR        F1=0.740±0.013  nDCG@10=0.940  XAIR@10=0.000
  [FIRE-AgentIR] RETRO-style      F1=0.716±0.014  nDCG@10=0.933  XAIR@10=0.000
  [FIRE-AgentIR] MEIRA-full       F1=0.826±0.011  nDCG@10=0.970  X

### **End of Paper-1a**